src/nupic/ntypes/Sdr.hpp

```diff
@@ -0,0 +1,273 @@
+ *    in parallel. This format is useful because it contains the location of
+ *    each true bit inside of the SDR's dimensional space.
+ *
+ *    Flat Index Format: A list of the indices of the true bits, indexed into
+ *    the the flattened SDR.  This format allows for quickly accessing all of
+ *    the true bits in the SDR.
+ *
+ * Flat Formats: This class uses C-order throughout, meaning that when iterating
+ * through the SDR, the last/right-most index changes fastest.
+ *
+ * Boolean Data Type: Although the documentation for this class may refer to the
+ * values in an SDR as bits they are in fact stored as bytes (of type signed
+ * character).
+ *
+ * Example usage:
+ *
+ *    // Make an SDR with 10,000 bits, arranged in a (100 x 100) gird.
+ *    SparseDistributedRepresentation X({100, 100});
+ *
+ *    // Assign data in any format.
+ *    X.setFlatIndex( mySparseData ); // Assigns mySparseData to SDR's value.
+ *    X.setDense( myDenseData );      // myDenseData overwrites mySparseData.
+ *
+ *    // Access data in different formats.
+ *    indices      = X.getFlatIndex(); // Calculated from myDenseData.
+ *    indicesAgain = X.getFlatIndex(); // Reuses previous result, no work done.
+ *
+ */
+class SparseDistributedRepresentation : public Serializable
+{
+public:
+    /**
+     * Create an uninitialized SDR.  The only way to initialize SDRs created by
+     * this method is by calling SDR.load().  Until SDR.load() is called this
+     * SDR is in an unusable state.
+     */
+    SparseDistributedRepresentation();
+
+    /**
+     * Create an SDR object.  Initially this SDR has no value set.
+     *
+     * @param dimensions A list of dimension sizes, defining the shape of the
+     * SDR.
+     */
+    SparseDistributedRepresentation(const vector<UInt> dimensions);
+
+    /**
+     * Initialize this SDR as a copy of the given SDR.  This is a deep copy
+     * meaning that all of the data in the given SDR is fully copied, and
+     * modifying either SDR will not affect the other SDR.
+     *
+     * @param value An SDR to replicate.
+     */
+    SparseDistributedRepresentation(const SparseDistributedRepresentation value);
+
+    /**
+     * @returns A list of dimensions of the SDR.
+     */
+    const vector<UInt> getDimensions() const
+        {return dimensions;};
+
+    /**
+     * @returns The total number of boolean values in the SDR.
+     */
+    const UInt getSize() const
+        {return size;};
+
+    /**
+     * Remove the value from this SDR.  Attempting to get the value of an empty
+     * SDR will raise an exception.
+     */
+    void clear();
+
+    /**
+     * Set all of the values in the SDR to false.  This method overwrites the
+     * SDRs current value.
+     */
+    void zero();
+
+    /**
+     * Assigns a new value to the SDR, overwritting the current value.
+     *
+     * @param value A dense array in a vector<char> to assign to the SDR.
+     */
+    void setDense( const vector<Byte> value );
```

Comment:
You'd like for this to store non-boolean values.   I've experimented in python with SDRs using non-boolean data, integers in my case.  I've found it adds a considerable amount of complexity (and that's in a dynamically typed language) for marginal gains in functionality.  It makes everything more complicated, both in the SDR class and in code calling the SDR class.

It also doesn't fit in with my concept of an SDR.  It complicates the mathematics and the practices of using SDRs.

I found that the most useful thing I did with the values in my SDR was in my TM implementation: to differentiate between bursting & predicted activations.  For this use-case I could have used binary SDRs because there are only 3 levels of activity (inactive, active-minicolumn-burst, active-predicted).  I simply don't see the need for a real valued bit in an SDR.

Reply:
I did not state this very well.  SDR's are boolean values.

What I am trying to figure out is how to deal with NON-SDR data that is passed between regions. Currently both SDR and non-SDR data is passed in the same way via the links.  We need to think this through.